lib/application/pages/home/progress/progress_page.dart

```diff
@@ -0,0 +1,222 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_hooks/flutter_hooks.dart';
+import 'package:hooks_riverpod/hooks_riverpod.dart';
+import 'package:layoutr/common_layout.dart';
+import 'package:memo/application/constants/dimensions.dart' as dimens;
+import 'package:memo/application/constants/strings.dart' as strings;
+import 'package:memo/application/theme/theme_controller.dart';
+import 'package:memo/application/view-models/home/progress_vm.dart';
+import 'package:memo/application/widgets/animatable_progress.dart';
+
+class ProgressPage extends HookWidget {
+  @override
+  Widget build(BuildContext context) {
+    final state = useProvider(progressVM.state);
+    if (state is LoadingProgressState) {
+      return const Center(child: CircularProgressIndicator());
+    }
+
+    final loadedState = state as LoadedProgressState;
+
+    final firstRowParallelContainers = Row(
+      crossAxisAlignment: CrossAxisAlignment.stretch,
+      children: [
+        Flexible(
+          child: _buildTotalTimeContainer(context, timeProgress: loadedState.timeProgress),
+        ),
+        context.horizontalBox(Spacing.medium),
+        Flexible(
+          child: _ProgressContainer(
+            title: _buildAlternateStyleTextBox(
+              context,
+              texts: [loadedState.completedMemosCount.toString()],
+            ),
+            description: strings.progressTotalMemos,
+          ),
+        ),
+      ],
+    );
+
+    final hardMemosProgress = _ProgressContainer(
+      leading: _buildCircularProgress(
+        context,
+        progressValue: loadedState.hardMemosPercentage,
+        centerLabel: strings.faceScreamingInFear,
+        semanticLabel: strings.progressHardMemosIndicatorLabel,
+      ),
+      title: _buildAlternateStyleTextBox(
+        context,
+        texts: [loadedState.readableHardMemosPercentage, strings.percentSymbol],
+      ),
+      description: strings.progressTotalHardMemos,
+    );
+
+    final mediumMemosProgress = _ProgressContainer(
+      leading: _buildCircularProgress(
+        context,
+        progressValue: loadedState.mediumMemosPercentage,
+        centerLabel: strings.expressionlessFace,
+        semanticLabel: strings.progressMediumMemosIndicatorLabel,
+      ),
+      title: _buildAlternateStyleTextBox(
+        context,
+        texts: [loadedState.readableMediumMemosPercentage, strings.percentSymbol],
+      ),
+      description: strings.progressTotalMediumMemos,
+    );
+
+    final easyMemosProgress = _ProgressContainer(
+      leading: _buildCircularProgress(
+        context,
+        progressValue: loadedState.easyMemosPercentage,
+        centerLabel: strings.squintingFaceWithTongue,
+        semanticLabel: strings.progressEasyMemosIndicatorLabel,
+      ),
+      title: _buildAlternateStyleTextBox(
+        context,
+        texts: [loadedState.readableEasyMemosPercentage, strings.percentSymbol],
+      ),
+      description: strings.progressTotalEasyMemos,
+    );
```

Comment:
All this progress components are very similar. Have you tried to create an abstract function to create them? Or the amount of parameters from this function makes the abstraction approach worst than the duplicate code?

Reply:
But nothing is duplicate here, they are simply passing arguments to a widget and two functions that build two other dependent widgets. If one would create a function, it would possibly take the same amount of "lines", as the arguments from each other are completely different (other than using `context` twice and `strings.percentSymbol` being shared among them all)